Add is_built_in() method to Resource

KoBeWi opened the pull request:
A built-in Resource can be recognized by its path, which may be:
- empty
- containing `::`
- starting with `local://`? idk, I've seen it in few places

There were such checks all over the codebase, but they were inconsistent, which led to some bugs, e.g. #45950 or partially #49086 (`_is_built_in_script()` is not checking for empty paths, so the issue isn't fully fixed).

To fix inconsistencies I added `is_built_in()` to resources. I replaced equivalent conditions with this method.

KoBeWi committed:
Add is_built_in() method to Resource

akien-mga reviewed (approved):
Let's merge, but maybe warrants a rebase first to be sure it still builds fine.

KoBeWi commented:
Rebased. I also replaced a few more conditions (that likely popped out since the PR was made).

I seen a few partial conditions, which only check for `"::"`, I didn't touch them as I wasn't sure.
There is also `String.is_resource_file()`, which I didn't replace too.

akien-mga commented:
> I seen a few partial conditions, which only check for `"::"`, I didn't touch them as I wasn't sure.

Would probably be worth reviewing in a follow-up PR, I guess the intent was to check if it's a builtin resource. What else would use `::`?

KoBeWi commented:
The ones left out are:
https://github.com/godotengine/godot/blob/78931aa040bb0d0586345f95cc982605c0720ba1/core/string/ustring.cpp#L4338-L4340
This is used in lots of places. Some of them (`res.get_path().is_resource_file()`) could be replaced by `is_built_in()`.

https://github.com/godotengine/godot/blob/78931aa040bb0d0586345f95cc982605c0720ba1/editor/dependency_editor.cpp#L174-L180
Doesn't look related to file paths.

https://github.com/godotengine/godot/blob/78931aa040bb0d0586345f95cc982605c0720ba1/editor/editor_file_system.cpp#L122-L127
No idea here.

https://github.com/godotengine/godot/blob/78931aa040bb0d0586345f95cc982605c0720ba1/editor/editor_node.cpp#L1200-L1208
It uses the index of "::", but the performance probably doesn't matter here. Could be replaced.

https://github.com/godotengine/godot/blob/78931aa040bb0d0586345f95cc982605c0720ba1/editor/editor_node.cpp#L2159-L2168
Same as above.

https://github.com/godotengine/godot/blob/78931aa040bb0d0586345f95cc982605c0720ba1/editor/filesystem_dock.cpp#L1245-L1249
Not sure here.

https://github.com/godotengine/godot/blob/78931aa040bb0d0586345f95cc982605c0720ba1/editor/debugger/editor_debugger_inspector.cpp#L152-L160
Could be replaced probably.

https://github.com/godotengine/godot/blob/78931aa040bb0d0586345f95cc982605c0720ba1/editor/plugins/script_editor_plugin.cpp#L722-L723
It checks `!path.is_resource_file()` later, so it confused me a bit.

akien-mga commented:
Thanks!

akien-mga commented:
Cherry-picked for 3.5.
